File: 2516-take-k-of-each-character-from-left-and-right/2516-take-k-of-each-character-from-left-and-right.py
class Solution:
    def takeCharacters(self, s: str, k: int) -> int:
        count, n = Counter(s), len(s)
        j = window = 0
        if count['a'] < k or count['b'] < k or count['c'] < k:
            return -1
        for i, ch in enumerate(s):
            count[ch] -= 1
            while count[ch] < k:
                count[s[j]] += 1
                j += 1
            window = max(window, i - j + 1)
        return n - window

File: 2516-take-k-of-each-character-from-left-and-right/2516-take-k-of-each-character-from-left-and-right.js
/**
 * @param {string} s
 * @param {number} k
 * @return {number}
 */
var takeCharacters = function(s, k) {
  const count = [0, 0, 0];
  const n = s.length;
  const a = 'a'.charCodeAt();
  let [l, r, window] = [0, 0, 0];
  
  for (let ch of s) count[ch.charCodeAt() - a]++;
  
  if (count.some((x) => x < k)) return -1;
  
  while (r < n) {
    const index = s.charCodeAt(r) - a;
    count[index]--;
    
    while (count[index] < k) count[s.charCodeAt(l++) - a]++;
    
    window = Math.max(window, r - l + 1);
    r++;
  }
  
  return n - window;
};